build-data-capturing-gradle-samples/capture-thermal-throttling/gradle-thermal-throttling.gradle

```diff
@@ -0,0 +1,139 @@
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.os.OperatingSystem
+
+import java.nio.charset.Charset
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+
+/**
+ * This Gradle script captures the thermal throttling level and adds it as a tag.
+ * Some parameters can be tweaked according to your build:
+ * - SAMPLING_INTERVAL_IN_SECONDS, frequency at which the capture command is run
+ * - THROTLLING_LEVEL, list of throttling levels and value ranges (to be compared with captured throttling average value)
+ *
+ * WARNINGS:
+ * - This is supported on MacOS only.
+ *
+ * This should be applied to the root project:
+ * <code> apply from: file('gradle-thermal-throttling.gradle') </code>
+ */
+
+def buildScanApi = project.extensions.findByName('buildScan')
+if (!buildScanApi) {
+  return
+}
+
+// Register Thermal throttling service
+def throttlingServiceProvider = gradle.sharedServices.registerIfAbsent("thermalThrottling", ThermalThrottlingService.class) {}
+
+// Start Thermal throttling service
+throttlingServiceProvider.get()
+
+// Process results on build completion
+buildScan.buildFinished { def result ->
+  throttlingServiceProvider.get().processResults(buildScanApi)
+}
+
+// Thermal Throttling service implementation
+abstract class ThermalThrottlingService implements BuildService<BuildServiceParameters.None>, AutoCloseable {
+
+  /**
+   * Sampling interval can be adjusted according to total build time.
+   */
+  def static final SAMPLING_INTERVAL_IN_SECONDS = 5
+
+  /**
+   * Throttling levels by throttling average value.
+   */
+  def static final THROTTLING_LEVEL =
+          [
+                  [level: "THROTTLING_HIGH", range: 0..40],
+                  [level: "THROTTLING_MEDIUM", range: 40..80],
+                  [level: "THROTTLING_LOW", range: 80..100]
+          ]
+
+  def static final COMMAND_ARGS = ["pmset", "-g", "therm"]
+  def static final COMMAND_OUTPUT_PARSING_PATTERN = /CPU_Speed_Limit\s+=\s+/
+
+  def final scheduler
+  def final samples
+
+  ThermalThrottlingService() {
+    scheduler = Executors.newScheduledThreadPool(1)
+    samples = new ConcurrentLinkedQueue<Integer>()
+    if (OperatingSystem.current().macOsX) {
```

Comment:
should we check this at the very beginning before registering the build service? this way absolutely no build logic is added on unsupported platforms?

Reply:
that's a good point, let me adjust this